packages/buidler-etherscan/src/index.ts

```diff
@@ -1,97 +1,269 @@
-
-import AbiEncoder from "./AbiEncoder";
-import { getDefaultEtherscanConfig } from "./config";
-import {
-  getVerificationStatus,
-  verifyContract,
-} from "./etherscan/EtherscanService";
-import { toRequest } from "./etherscan/EtherscanVerifyContractRequest";
-import { getLongVersion } from "./solc/SolcVersions";
-import { EtherscanConfig } from "./types";
-
-task("verify-contract", "Verifies contract on etherscan")
-  .addParam("contractName", "Name of the deployed contract")
-  .addParam("address", "Deployed address of smart contract")
-  .addOptionalParam(
-    "libraries",
-    'Stringified JSON object in format of {library1: "0x2956356cd2a2bf3202f771f50d3d14a367b48071"}'
-  )
-  .addOptionalVariadicPositionalParam(
-    "constructorArguments",
-    "arguments for contract constructor",
-    []
-  )
-  .setAction(
-    async (
-      taskArgs: {
-        contractName: string;
-        address: string;
-        libraries: string;
-        source: string;
-        constructorArguments: string[];
-      },
-      { config, run }
-    ) => {
-      const etherscan: EtherscanConfig = getDefaultEtherscanConfig(config);
-
-      if (etherscan.apiKey === undefined || etherscan.apiKey.trim() === "") {
+import { NomicLabsBuidlerPluginError } from "@nomiclabs/buidler/plugins";
+import { ActionType, Artifact } from "@nomiclabs/buidler/types";
+import SemverRange from "semver/classes/range";
+
+import { pluginName } from "./pluginContext";
+
+interface VerificationArgs {
+  address: string;
+  constructorArguments: string[];
+  // Filename of constructor arguments module.
+  constructorArgs?: string;
+}
+
+const verify: ActionType<VerificationArgs> = async (
+  {
+    address,
+    constructorArguments: constructorArgsList,
+    constructorArgs: constructorArgsModule,
+  },
+  { config, network, run }
+) => {
+  const { getDefaultEtherscanConfig } = await import("./config");
+  const etherscan = getDefaultEtherscanConfig(config);
+
+  if (
+    etherscan.apiKey === undefined ||
+    etherscan.apiKey === null ||
+    etherscan.apiKey.trim() === ""
+  ) {
+    throw new NomicLabsBuidlerPluginError(
+      pluginName,
+      `Please provide an Etherscan API token via buidler config.
+E.g.: { [...], etherscan: { apiKey: 'an API key' }, [...] }
+See https://etherscan.io/apis`
+    );
+  }
+
+  if (network.name === "buidlerevm") {
+    throw new NomicLabsBuidlerPluginError(
+      pluginName,
+      `The selected network is ${network.name}. Please select a network supported by Etherscan.`
+    );
+  }
+
+  const { isAddress } = await import("@ethersproject/address");
+  if (!isAddress(address)) {
+    throw new NomicLabsBuidlerPluginError(
+      pluginName,
+      `${address} is an invalid address.`
+    );
+  }
+
+  const { getVersionNumber, inferSolcVersion, InferralType } = await import(
+    "./solc/version"
+  );
+  const solcVersionConfig = getVersionNumber(config.solc.version);
+
+  const supportedSolcVersionRange = new SemverRange(">=0.4.11");
```

Comment:
I wouldn't implement this validation. 0.4.11 is super old.

I'd just leave a comment in the source code explaining that this will fail with solc < 0.4.11.